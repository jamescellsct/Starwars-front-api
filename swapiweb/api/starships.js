import express from "express";

const router = express.Router();

var starships = require("./fakeData/db.json").starships;

router.get("/starships", function(req, res) {
  return res.status(200).json({
    count: starships.count,
    results: starships.results[0]
  });
});

router.get("/starships/:page", function(req, res) {
  let { page } = req.params;
  page = page.slice(5, 6) - 1;

  if (!starships.results[page]) {
    return res.status(404).send({
      error: {
        message: "Page Not Found!"
      }
    });
  }

  return res.status(200).json({
    count: starships.count,
    results: starships.results[page]
  });
});

export default router;
