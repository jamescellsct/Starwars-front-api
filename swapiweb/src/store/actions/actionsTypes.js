//Peoples Actions Types
const FETCH_PEOPLE_STARTED = "FETCH_PEOPLE_STARTED";
const FETCH_PEOPLE_SUCCEEDED = "FETCH_PEOPLE_SUCCEEDED";
const FETCH_PEOPLE_FAILED = "FETCH_PEOPLE_FAILED";
const UPDATE_PEOPLE_IS_FETCHING = "UPDATE_PEOPLE_IS_FETCHING";
const UPDATE_PEOPLE_SUCCEEDED = "UPDATE_PEOPLE_SUCCEEDED";
const UPDATE_PEOPLE_FAILED = "UPDATE_PEOPLE_FAILED";
export const peoplesAction = {
  FETCH_PEOPLE_STARTED,
  FETCH_PEOPLE_SUCCEEDED,
  FETCH_PEOPLE_FAILED,
  UPDATE_PEOPLE_IS_FETCHING,
  UPDATE_PEOPLE_SUCCEEDED,
  UPDATE_PEOPLE_FAILED
};

//Planets Actions Types
const FETCH_PLANETS_STARTED = "FETCH_PLANETS_STARTED";
const FETCH_PLANETS_SUCCEEDED = "FETCH_PLANETS_SUCCEEDED";
const FETCH_PLANETS_FAILED = "FETCH_PLANETS_FAILED";
const UPDATE_PLANETS_IS_FETCHING = "UPDATE_PLANETS_IS_FETCHING";
const UPDATE_PLANETS_SUCCEEDED = "UPDATE_PLANETS_SUCCEEDED";
const UPDATE_PLANETS_FAILED = "UPDATE_PLANETS_FAILED";
export const planetsAction = {
  FETCH_PLANETS_STARTED,
  FETCH_PLANETS_SUCCEEDED,
  FETCH_PLANETS_FAILED,
  UPDATE_PLANETS_IS_FETCHING,
  UPDATE_PLANETS_SUCCEEDED,
  UPDATE_PLANETS_FAILED
};

//Starships Actions Types
const FETCH_STARSHIPS_STARTED = "FETCH_STARSHIPS_STARTED";
const FETCH_STARSHIPS_SUCCEEDED = "FETCH_STARSHIPS_SUCCEEDED";
const FETCH_STARSHIPS_FAILED = "FETCH_STARSHIPS_FAILED";
const UPDATE_STARSHIPS_IS_FETCHING = "UPDATE_STARSHIPS_IS_FETCHING";
const UPDATE_STARSHIPS_SUCCEEDED = "UPDATE_STARSHIPS_SUCCEEDED";
const UPDATE_STARSHIPS_FAILED = "UPDATE_STARSHIPS_FAILED";
export const starshipsAction = {
  FETCH_STARSHIPS_STARTED,
  FETCH_STARSHIPS_SUCCEEDED,
  FETCH_STARSHIPS_FAILED,
  UPDATE_STARSHIPS_IS_FETCHING,
  UPDATE_STARSHIPS_SUCCEEDED,
  UPDATE_STARSHIPS_FAILED
};

//Vehicles Actions Types
const FETCH_VEHICLES_STARTED = "FETCH_VEHICLES_STARTED";
const FETCH_VEHICLES_SUCCEEDED = "FETCH_VEHICLES_SUCCEEDED";
const FETCH_VEHICLES_FAILED = "FETCH_VEHICLES_FAILED";
const UPDATE_VEHICLES_IS_FETCHING = "UPDATE_VEHICLES_IS_FETCHING";
const UPDATE_VEHICLES_SUCCEEDED = "UPDATE_VEHICLES_SUCCEEDED";
const UPDATE_VEHICLES_FAILED = "UPDATE_VEHICLES_FAILED";
export const vehiclesAction = {
  FETCH_VEHICLES_STARTED,
  FETCH_VEHICLES_SUCCEEDED,
  FETCH_VEHICLES_FAILED,
  UPDATE_VEHICLES_IS_FETCHING,
  UPDATE_VEHICLES_SUCCEEDED,
  UPDATE_VEHICLES_FAILED
};

//Pagination Action Types
const PAGE_SELECTED = "PAGE_SELECTED";
export const paginationAction = {
  PAGE_SELECTED
};
