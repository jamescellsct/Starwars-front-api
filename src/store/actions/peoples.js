import { peoplesAction } from "./actionsTypes";
import { getPeople } from "../../services/api/peopleService";
import { handleResulstWithId } from "../../utils/handleResultsWithId";

const fetchPeopleStarted = () => ({
  type: peoplesAction.FETCH_PEOPLE_STARTED
});

const fetchPeopleSucceeded = people => ({
  type: peoplesAction.FETCH_PEOPLE_SUCCEEDED,
  people
});

const fetchPeopleFailed = error => ({
  type: peoplesAction.FETCH_PEOPLE_FAILED,
  error
});

export const fetchPeople = () => dispatch => {
  dispatch(fetchPeopleStarted());

  return getPeople()
    .then(response => {
      const { results, count } = response;
      dispatch(
        fetchPeopleSucceeded({
          page: 1,
          results: handleResulstWithId(results, 1),
          pageCount: Math.ceil(count / 10)
        })
      );
    })
    .catch(error => {
      dispatch(fetchPeopleFailed(error));
    });
};

const updatePeopleIsFetching = () => ({
  type: peoplesAction.UPDATE_PEOPLE_IS_FETCHING
});

const updatePeopleSucceeded = people => ({
  type: peoplesAction.UPDATE_PEOPLE_SUCCEEDED,
  people
});

const updatePeopleFailed = error => ({
  type: peoplesAction.UPDATE_PEOPLE_FAILED,
  error
});

export const updatePeople = page => dispatch => {
  dispatch(updatePeopleIsFetching());

  getPeople(page)
    .then(response => {
      const { results, count } = response;
      dispatch(
        updatePeopleSucceeded({
          page,
          results: handleResulstWithId(results, page),
          pageCount: Math.ceil(count / 10)
        })
      );
    })
    .catch(error => {
      dispatch(updatePeopleFailed(error));
    });
};
