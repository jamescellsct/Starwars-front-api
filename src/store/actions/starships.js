import { starshipsAction } from "./actionsTypes";
import { getStarships } from "../../services/api/starshipsService";
import { handleResulstWithId } from "../../utils/handleResultsWithId";

const fetchStarshipsStarted = () => ({
  type: starshipsAction.FETCH_STARSHIPS_STARTED
});

const fetchStarshipsSucceeded = starships => ({
  type: starshipsAction.FETCH_STARSHIPS_SUCCEEDED,
  starships
});

const fetchStarshipsFailed = error => ({
  type: starshipsAction.FETCH_STARSHIPS_FAILED,
  error
});

export const fetchStarships = () => dispatch => {
  dispatch(fetchStarshipsStarted());

  return getStarships()
    .then(response => {
      const { results, count } = response;
      dispatch(
        fetchStarshipsSucceeded({
          page: 1,
          results: handleResulstWithId(results, 1),
          pageCount: Math.ceil(count / 10)
        })
      );
    })
    .catch(error => {
      dispatch(fetchStarshipsFailed(error));
    });
};

const updateStarshipsIsFetching = () => ({
  type: starshipsAction.UPDATE_STARSHIPS_IS_FETCHING
});

const updateStarshipsSucceeded = starships => ({
  type: starshipsAction.UPDATE_STARSHIPS_SUCCEEDED,
  starships
});

const updateStarshipsFailed = error => ({
  type: starshipsAction.UPDATE_STARSHIPS_FAILED,
  error
});

export const updateStarships = page => dispatch => {
  dispatch(updateStarshipsIsFetching());
  return getStarships(page)
    .then(response => {
      const { results, count } = response;
      dispatch(
        updateStarshipsSucceeded({
          page,
          results: handleResulstWithId(results, page),
          pageCount: Math.ceil(count / 10)
        })
      );
    })
    .catch(error => {
      dispatch(updateStarshipsFailed(error));
    });
};
