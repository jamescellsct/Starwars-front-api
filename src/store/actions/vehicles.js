import { vehiclesAction } from "./actionsTypes";
import { getVehicles } from "../../services/api/vehiclesService";
import { handleResulstWithId } from "../../utils/handleResultsWithId";

const fetchVehiclesStarted = () => ({
  type: vehiclesAction.FETCH_VEHICLES_STARTED
});

const fetchVehiclesSucceeded = vehicles => ({
  type: vehiclesAction.FETCH_VEHICLES_SUCCEEDED,
  vehicles
});

const fetchVehiclesFailed = error => ({
  type: vehiclesAction.FETCH_VEHICLES_FAILED,
  error
});

export const fetchVehicles = () => dispatch => {
  dispatch(fetchVehiclesStarted());

  return getVehicles()
    .then(response => {
      const { results, count } = response;
      dispatch(
        fetchVehiclesSucceeded({
          page: 1,
          results: handleResulstWithId(results, 1),
          pageCount: Math.ceil(count / 10)
        })
      );
    })
    .catch(error => {
      dispatch(fetchVehiclesFailed(error));
    });
};

const updateVehiclesIsFetching = () => ({
  type: vehiclesAction.UPDATE_VEHICLES_IS_FETCHING
});

const updateVehiclesSucceeded = vehicles => ({
  type: vehiclesAction.UPDATE_VEHICLES_SUCCEEDED,
  vehicles
});

const updateVehiclesFailed = error => ({
  type: vehiclesAction.UPDATE_VEHICLES_FAILED,
  error
});

export const updateVehicles = page => dispatch => {
  dispatch(updateVehiclesIsFetching());
  return getVehicles(page)
    .then(response => {
      const { results, count } = response;
      dispatch(
        updateVehiclesSucceeded({
          page,
          results: handleResulstWithId(results, page),
          pageCount: Math.ceil(count / 10)
        })
      );
    })
    .catch(error => {
      dispatch(updateVehiclesFailed(error));
    });
};
