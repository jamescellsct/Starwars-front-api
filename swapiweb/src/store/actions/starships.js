import { starshipsAction } from "./actionsTypes";
import { swapiRequest } from "../../services/swapi_connect";

const fetchStarshipsStarted = () => ({
  type: starshipsAction.FETCH_STARSHIPS_STARTED
});

const fetchStarshipsSucceeded = starships => ({
  type: starshipsAction.FETCH_STARSHIPS_SUCCEEDED,
  starships
});

const fetchStarshipsFailed = error => ({
  type: starshipsAction.FETCH_STARSHIPS_FAILED,
  error
});

export const fetchStarships = () => dispatch => {
  dispatch(fetchStarshipsStarted());

  swapiRequest(null, null, "starships")
    .then(response => {
      dispatch(
        fetchStarshipsSucceeded({
          results: response.results,
          total: response.count
        })
      );
    })
    .catch(error => {
      dispatch(fetchStarshipsFailed(error));
    });
};
