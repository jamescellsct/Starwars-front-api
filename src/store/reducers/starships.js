import { starshipsAction } from "../actions/actionsTypes";

export default function starships(state = null, action) {
  switch (action.type) {
    case starshipsAction.FETCH_STARSHIPS_STARTED:
      return Object.assign({}, state, {
        isFetching: true
      });

    case starshipsAction.FETCH_STARSHIPS_SUCCEEDED:
      return Object.assign({}, state, {
        pages: [
          ...(state.pages || []),
          { page: action.starships.page, list: action.starships.results }
        ],
        pageCount: action.starships.pageCount,
        isFetching: false
      });

    case starshipsAction.FETCH_STARSHIPS_FAILED:
      return Object.assign({}, state, {
        isFetching: false,
        error: action.error
      });

    case starshipsAction.UPDATE_STARSHIPS_IS_FETCHING:
      return Object.assign({}, state, {
        isFetching: true
      });
    case starshipsAction.UPDATE_STARSHIPS_SUCCEEDED:
      return Object.assign({}, state, {
        pages: [
          ...(state.pages || []),
          { page: action.starships.page, list: action.starships.results }
        ],
        pageCount: action.starships.pageCount,
        isFetching: false
      });

    case starshipsAction.UPDATE_STARSHIPS_FAILED:
      return Object.assign({}, state, {
        isFetching: false,
        error: action.error
      });

    default:
      return state;
  }
}
