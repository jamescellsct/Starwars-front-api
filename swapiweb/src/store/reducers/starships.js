import { starshipsAction } from "../actions/actionsTypes";

export default function starships(state = null, action) {
  switch (action.type) {
    case starshipsAction.FETCH_STARSHIPS_STARTED:
      return Object.assign({}, state, {
        isFetching: true
      });

    case starshipsAction.FETCH_STARSHIPS_SUCCEEDED:
      return Object.assign({}, state, {
        list: [...(state.list || []), ...action.starships.results],
        isFetching: false
      });

    case starshipsAction.FETCH_STARSHIPS_FAILED:
      return Object.assign({}, state, {
        isFetching: false,
        error: action.error
      });

    default:
      return state;
  }
}
