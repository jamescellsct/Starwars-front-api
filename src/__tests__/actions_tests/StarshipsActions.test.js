import thunk from "redux-thunk";
import configureMockStore from "redux-mock-store";
import nock from "nock";

import { starshipsAction } from "../../store/actions/actionsTypes";
import { fetchStarships, updateStarships } from "../../store/actions/starships";
import BASE_URL from "./baseUrl";

const midlewares = [thunk];
const mockStore = configureMockStore(midlewares);

const mockResults = [
  {
    name: "Dale Jackson",
    avatar: "http://i.pravatar.cc/40?img=0.7474128780010723",
    model: "so",
    manufacturer: "vajcez",
    length: 191,
    max_atmosphering_speed: 205,
    passengers: 1473,
    cargo_capacity: 5579622,
    consumables: 1,
    starship_class: "rukvu",
    idItem: 1
  },
  {
    name: "Sylvia Barton",
    avatar: "http://i.pravatar.cc/40?img=0.7575125938238629",
    model: "ejvoj",
    manufacturer: "nubodivas",
    length: 17,
    max_atmosphering_speed: 522,
    passengers: 96,
    cargo_capacity: 2861477,
    consumables: 5,
    starship_class: "lu",
    idItem: 2
  }
];

const mockUpdateResults = [
  {
    name: "Updating Jackson",
    avatar: "http://i.pravatar.cc/40?img=0.7474128780010723",
    model: "so",
    manufacturer: "vajcez",
    length: 191,
    max_atmosphering_speed: 205,
    passengers: 1473,
    cargo_capacity: 5579622,
    consumables: 1,
    starship_class: "rukvu",
    idItem: 1
  },
  {
    name: "Updating Barton",
    avatar: "http://i.pravatar.cc/40?img=0.7575125938238629",
    model: "ejvoj",
    manufacturer: "nubodivas",
    length: 17,
    max_atmosphering_speed: 522,
    passengers: 96,
    cargo_capacity: 2861477,
    consumables: 5,
    starship_class: "lu",
    idItem: 2
  }
];

describe("Testing Starships Action", () => {
  it("Shold create an action to fetching starships data", () => {
    afterEach(() => {
      nock.cleanAll();
    });

    nock(BASE_URL)
      .defaultReplyHeaders({ "access-control-allow-origin": "*" })
      .get("/starships")
      .reply(200, {
        count: 20,
        results: mockResults
      });

    const expectedAction = [
      { type: starshipsAction.FETCH_STARSHIPS_STARTED },
      {
        type: starshipsAction.FETCH_STARSHIPS_SUCCEEDED,
        starships: {
          page: 1,
          results: mockResults,
          pageCount: 2
        }
      }
    ];

    const store = mockStore({ starships: {} });
    return store.dispatch(fetchStarships()).then(() => {
      expect(store.getActions()).toEqual(expectedAction);
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it("Should return ERROR in fetching starships data", () => {
    afterEach(() => {
      nock.cleanAll();
    });

    nock(BASE_URL)
      .defaultReplyHeaders({ "access-control-allow-origin": "*" })
      .get("/starships")
      .reply(400, {});

    const expectedAction = [
      { type: starshipsAction.FETCH_STARSHIPS_STARTED },
      {
        type: starshipsAction.FETCH_STARSHIPS_FAILED,
        error: new Error("Request failed with status code 400")
      }
    ];

    const store = mockStore({ starships: {} });
    return store.dispatch(fetchStarships()).then(() => {
      expect(store.getActions()).toEqual(expectedAction);
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it("Should create an action to update starships data", () => {
    afterEach(() => {
      nock.cleanAll();
    });

    nock(BASE_URL)
      .defaultReplyHeaders({ "access-control-allow-origin": "*" })
      .get("/starships/page=1")
      .reply(200, {
        count: 20,
        results: mockUpdateResults
      });

    const expectedAction = [
      { type: starshipsAction.UPDATE_STARSHIPS_IS_FETCHING },
      {
        type: starshipsAction.UPDATE_STARSHIPS_SUCCEEDED,
        starships: {
          page: 1,
          results: mockUpdateResults,
          pageCount: 2
        }
      }
    ];

    const store = mockStore({ starships: {} });
    return store.dispatch(updateStarships(1)).then(() => {
      expect(store.getActions()).toEqual(expectedAction);
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it("Should return ERROR on updating starships data", () => {
    afterEach(() => {
      nock.cleanAll();
    });

    nock(BASE_URL)
      .defaultReplyHeaders({ "access-control-allow-origin": "*" })
      .get("/starships/page=1")
      .reply(400, {});

    const expectedAction = [
      { type: starshipsAction.UPDATE_STARSHIPS_IS_FETCHING },
      {
        type: starshipsAction.UPDATE_STARSHIPS_FAILED,
        error: new Error("Request failed with status code 400")
      }
    ];

    const store = mockStore({ starships: {} });
    return store.dispatch(updateStarships(1)).then(() => {
      expect(store.getActions()).toEqual(expectedAction);
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
