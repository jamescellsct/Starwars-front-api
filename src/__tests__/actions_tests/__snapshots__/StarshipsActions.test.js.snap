// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Starships Action Shold create an action to fetching starships data 1`] = `
Array [
  Object {
    "type": "FETCH_STARSHIPS_STARTED",
  },
  Object {
    "starships": Object {
      "page": 1,
      "pageCount": 2,
      "results": Array [
        Object {
          "avatar": "http://i.pravatar.cc/40?img=0.7474128780010723",
          "cargo_capacity": 5579622,
          "consumables": 1,
          "idItem": 1,
          "length": 191,
          "manufacturer": "vajcez",
          "max_atmosphering_speed": 205,
          "model": "so",
          "name": "Dale Jackson",
          "passengers": 1473,
          "starship_class": "rukvu",
        },
        Object {
          "avatar": "http://i.pravatar.cc/40?img=0.7575125938238629",
          "cargo_capacity": 2861477,
          "consumables": 5,
          "idItem": 2,
          "length": 17,
          "manufacturer": "nubodivas",
          "max_atmosphering_speed": 522,
          "model": "ejvoj",
          "name": "Sylvia Barton",
          "passengers": 96,
          "starship_class": "lu",
        },
      ],
    },
    "type": "FETCH_STARSHIPS_SUCCEEDED",
  },
]
`;

exports[`Testing Starships Action Should create an action to update starships data 1`] = `
Array [
  Object {
    "type": "UPDATE_STARSHIPS_IS_FETCHING",
  },
  Object {
    "starships": Object {
      "page": 1,
      "pageCount": 2,
      "results": Array [
        Object {
          "avatar": "http://i.pravatar.cc/40?img=0.7474128780010723",
          "cargo_capacity": 5579622,
          "consumables": 1,
          "idItem": 1,
          "length": 191,
          "manufacturer": "vajcez",
          "max_atmosphering_speed": 205,
          "model": "so",
          "name": "Updating Jackson",
          "passengers": 1473,
          "starship_class": "rukvu",
        },
        Object {
          "avatar": "http://i.pravatar.cc/40?img=0.7575125938238629",
          "cargo_capacity": 2861477,
          "consumables": 5,
          "idItem": 2,
          "length": 17,
          "manufacturer": "nubodivas",
          "max_atmosphering_speed": 522,
          "model": "ejvoj",
          "name": "Updating Barton",
          "passengers": 96,
          "starship_class": "lu",
        },
      ],
    },
    "type": "UPDATE_STARSHIPS_SUCCEEDED",
  },
]
`;

exports[`Testing Starships Action Should return ERROR in fetching starships data 1`] = `
Array [
  Object {
    "type": "FETCH_STARSHIPS_STARTED",
  },
  Object {
    "error": [Error: Request failed with status code 400],
    "type": "FETCH_STARSHIPS_FAILED",
  },
]
`;

exports[`Testing Starships Action Should return ERROR on updating starships data 1`] = `
Array [
  Object {
    "type": "UPDATE_STARSHIPS_IS_FETCHING",
  },
  Object {
    "error": [Error: Request failed with status code 400],
    "type": "UPDATE_STARSHIPS_FAILED",
  },
]
`;
